pkg/monitor/disruption_backend_sampler.go

```diff
@@ -0,0 +1,349 @@
+		path:                  path,
+	}
+}
+
+func NewRouteBackend(namespace, name, disruptionBackendName, path string, connectionType BackendConnectionType) *BackendSampler {
+	return &BackendSampler{
+		connectionType:        connectionType,
+		locator:               LocateRouteForDisruptionCheck(namespace, name, disruptionBackendName, connectionType),
+		disruptionBackendName: disruptionBackendName,
+		path:                  path,
+		routeCoordinates: &routeCoordinates{
+			namespace: namespace,
+			name:      name,
+		},
+	}
+}
+
+func (b *BackendSampler) WithHost(host string) *BackendSampler {
+	b.host = host
+	return b
+}
+
+func (b *BackendSampler) WithBearerTokenAuth(token, tokenFile string) *BackendSampler {
+	b.bearerToken = token
+	b.bearerTokenFile = tokenFile
+	return b
+}
+
+func (b *BackendSampler) WithTLSConfig(tlsConfig *tls.Config) *BackendSampler {
+	b.tlsConfig = tlsConfig
+	return b
+}
+
+func (b *BackendSampler) WithExpectedBody(expectedBody string) *BackendSampler {
+	b.expect = expectedBody
+	return b
+}
+
+func (b *BackendSampler) WithExpectedBodyRegex(expectedBodyRegex string) *BackendSampler {
+	b.expectRegexp = regexp.MustCompile(expectedBodyRegex)
+	return b
+}
+
+func (b *BackendSampler) SetEventRecorder(eventRecorder events.EventRecorder) {
+	b.eventRecorder = eventRecorder
+}
+
+func (b *BackendSampler) BodyMatches(body []byte) error {
+	switch {
+	case len(b.expect) != 0 && !bytes.Contains(body, []byte(b.expect)):
+		return fmt.Errorf("response did not contain the correct body contents: %q", string(body))
+	case b.expectRegexp != nil && !b.expectRegexp.MatchString(string(body)):
+		return fmt.Errorf("response did not contain the correct body contents: %q", string(body))
+	}
+
+	return nil
+}
+
+func (b *BackendSampler) GetDisruptionBackendName() string {
+	return b.disruptionBackendName
+}
+
+func (b *BackendSampler) GetLocator() string {
+	return b.locator
+}
+
+func (b *BackendSampler) GetConnectionType() BackendConnectionType {
+	return b.connectionType
+}
+
+func (b *BackendSampler) getTimeout() time.Duration {
+	if b.timeout == nil {
+		return 1 * time.Second
+	}
+	return *b.timeout
+}
+
+func (b *BackendSampler) GetURL() (string, error) {
+	var hostErr error
+	b.initializeHost.Do(func() {
+		if len(b.host) > 0 {
+			return
+		}
+		if b.routeCoordinates == nil {
+			hostErr = fmt.Errorf("no route coordinates to lookup host")
+			return
+		}
+		config, err := framework.LoadConfig()
+		if err != nil {
+			hostErr = err
+			return
+		}
+		client, err := routeclientset.NewForConfig(config)
+		if err != nil {
+			hostErr = err
+			return
+		}
+		route, err := client.RouteV1().Routes(b.routeCoordinates.namespace).Get(context.Background(), b.routeCoordinates.name, metav1.GetOptions{})
+		if err != nil {
+			hostErr = err
+			return
+		}
+		for _, ingress := range route.Status.Ingress {
+			if len(ingress.Host) > 0 {
+				b.host = fmt.Sprintf("https://%s", ingress.Host)
+				break
+			}
+		}
+	})
+	if hostErr != nil {
```

Comment:
may want to persist this to report the same error from all callers.  I need to read the "happens-before" semantic on .Do